File: Add Two Numbers.cpp
#include <bits/stdc++.h>
using namespace std;
// Definition for singly-linked list.
#include <cstddef>
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
    vector<int> d1;
    vector<int> d2;
    while (l1 != NULL) {
      d1.push_back(l1->val);
      l1 = l1->next;
    }
    while (l2 != NULL) {
      d2.push_back(l2->val);
      l2 = l2->next;
    }
    reverse(d1.begin(), d1.end());
    reverse(d2.begin(), d2.end());
    int idx1 = d1.size() - 1;
    int idx2 = d2.size() - 1;
    vector<int> sum(max(d1.size(), d2.size()) + 1);
    int idx3 = sum.size() - 1;
    int carry = 0;
    while (idx1 >= 0 || idx2 >= 0) {

      if (idx1 >= 0 && idx2 >= 0) {

        sum[idx3] = ((d1[idx1] + d2[idx2] + carry) % 10);
        carry = (d1[idx1] + d2[idx2] + carry) / 10;

      } else if (idx1) {
        sum[idx3] = (d1[idx1] + carry) % 10;
        carry = (d1[idx1] + carry) / 10;
      } else if (idx2) {

        sum[idx3] = (d2[idx2] + carry) % 10;
        carry = (d2[idx2] + carry) / 10;
      }

      idx1--;
      idx2--;
      idx3--;
    }

    int start = 0;

    while (start < sum.size() && sum[start] == 0)
      start++;

    if (start == sum.size()) {
      ListNode *x = new ListNode(0);
      return x;
    }

    int idx = sum.size() - 1;
    ListNode *head = NULL;
    ListNode *current = NULL;

    while (idx >= start) {
      if (head == NULL) {
        head = new ListNode(sum[idx]);
        head->next = NULL;
        current = head;
      } else {
        ListNode nw = ListNode(sum[idx]);
        nw.next = NULL;
        current->next = &nw;
        current = &nw;
      }
      idx--;
    }
    return head;
  }
};

int main() {
  ListNode x = ListNode(0);
  ListNode y = ListNode(1);
  x.next = &y;
  ListNode b = ListNode(7);
  ListNode c = ListNode(8);
  b.next = &c;
  ListNode *head = Solution().addTwoNumbers(&x, &b);

  while (head != NULL) {
     int x = head->val;
     cout<<x;
    head = head->next;
  }
  cout << endl;
}

File: Add Two Numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        d1 = []
        d2 = []

        while l1 != None:
            d1.append(l1.val)
            l1 = l1.next

        while l2 != None:
            d2.append(l2.val)
            l2 = l2.next

        d1.reverse()
        d2.reverse()

        sm = [0] * ((max(len(d1), len(d2))) + 1)
        idx1 = len(d1) - 1
        idx2 = len(d2) - 1
        idx3 = len(sm) - 1
        carry = 0

        while idx1 >= 0 or idx2 >= 0:

            if idx1 >= 0 and idx2 >= 0:
                sm[idx3] = (d1[idx1] + d2[idx2] + carry) % 10
                carry = (d1[idx1] + d2[idx2] + carry) // 10

            elif idx1 >= 0:
                sm[idx3] = (d1[idx1] + carry) % 10
                carry = (d1[idx1] + carry) // 10

            elif idx2 >= 0:
                sm[idx3] = (d2[idx2] + carry) % 10
                carry = (d2[idx2] + carry) // 10

            idx1 -= 1
            idx2 -= 1
            idx3 -= 1

        sm [idx3] = carry

        start = 0;

        while start < len(sm) and sm[start] == 0:
            start +=1

        if start == len(sm):
            return ListNode(0)

        idx = len(sm) - 1
        head = ListNode(sm[idx])
        current = head
        idx-=1

        while idx >= start:
            nw = ListNode(sm[idx])
            current.next = nw
            current = nw
            idx-=1

        return head


def main():
    l1 = convert([9,9,9,9,9,9,9])
    l2 = convert([9,9,9,9])
    head = Solution().addTwoNumbers(l1 , l2)

    while head != None:
        print(head.val)
        head = head.next


def convert(lst : list):
    head = ListNode(lst[0])
    current = head
    for i in range(1 , len(lst)):
        nw = ListNode(lst[i])
        current.next = nw
        current = nw

    return head

main()
